ddl/partition.go

```diff
@@ -1099,29 +1099,34 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
 		}
 	}
 
+	newIDs := make([]int64, len(oldIDs))
 	if d.infoHandle != nil && d.infoHandle.IsValid() {
 		bundles := make([]*placement.Bundle, 0, len(oldIDs))
 
-		yoldIDs := make([]int64, 0, len(oldIDs))
-		newIDs := make([]int64, 0, len(oldIDs))
 		for i, oldID := range oldIDs {
 			oldBundle, ok := d.infoHandle.Get().BundleByName(placement.GroupID(oldID))
 			if ok && !oldBundle.IsEmpty() {
-				yoldIDs = append(yoldIDs, oldID)
-				newIDs = append(newIDs, newPartitions[i].ID)
+				newIDs[i] = newPartitions[i].ID
 				bundles = append(bundles, placement.BuildPlacementDropBundle(oldID))
 				bundles = append(bundles, placement.BuildPlacementCopyBundle(oldBundle, newPartitions[i].ID))
+			} else {
+				newIDs[i] = -1
 			}
 		}
-		job.CtxVars = []interface{}{yoldIDs, newIDs}
+		job.CtxVars = []interface{}{oldIDs, newIDs}
 
 		err = infosync.PutRuleBundles(nil, bundles)
 		if err != nil {
 			job.State = model.JobStateCancelled
 			return ver, errors.Wrapf(err, "failed to notify PD the placement rules")
 		}
+	} else {
+		for i := range oldIDs {
+			newIDs[i] = -1
```

Comment:
`newIDs` should not only depend on the implementation of placement rules. By iteral, it means the new partition IDs.

Reply:
I knew. If you think always fetching rules are acceptable, or adding new fields to SchemaDiff is acceptable, we can do that.

Reply:
I think it's fine to always fetch rules from PD.

Reply:
Just one more time: is it acceptable to introduce, in the worst case, `2*x` RTT of TiDB/PD latency for the truncation of `x` partitions, even if the user is not using placement rules at all? Give me a `Yes`, and I will modify to always fetch rules.

Reply:
Why 2*x RTT? Deleting placement rules from the cache doesn't need fetching rules.

Reply:
Oh, yes. So there is constant factor `2`.

Reply:
Then I think it's fine. Please modify to always fetch rules.

Reply:
Done